Fix SparsePauliOp.is_unitary() to respect input tolerance values

littlebullGit opened the pull request:
Fixes #14107

### Summary

This PR fixes an issue where `SparsePauliOp.is_unitary()` was not properly respecting the input tolerance values when checking if an operator is unitary. The method was not passing the tolerance parameters to the internal [simplify()] call.

### Details and comments

The main changes in this PR are:

1. Modified `SparsePauliOp.is_unitary()` to pass the atol and rtol parameters to the [simplify()] method
2. Added a test case that verifies the fix works as expected by:
   - Creating a matrix that's approximately unitary but not exactly
   - Verifying it's considered unitary with appropriate tolerances
   - Verifying it's not considered unitary with very tight tolerances

### Testing
- [x] I have added the tests to cover my changes.
- [x] All tests pass locally with my changes
- [x] I have read the CONTRIBUTING document.

### Release Notes

Fixed an issue where `SparsePauliOp.is_unitary()` was not respecting the input tolerance values when checking if an operator is unitary.

littlebullGit committed:
Fix SparsePauliOp.is_unitary() to respect input tolerance values

Fixes #14107 by passing the tolerance parameters from is_unitary() to simplify().
This ensures that the tolerance values are properly respected when checking
if an operator is unitary.

The change includes a test case that verifies the fix works as expected.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=14676) <br/>All committers have signed the CLA.

Cryoris commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -449,7 +449,7 @@ def is_unitary(self, atol: float | None = None, rtol: float | None = None) -> bo
 
         # Compose with adjoint
-        val = self.compose(self.adjoint()).simplify()
+        val = self.compose(self.adjoint()).simplify(atol=atol, rtol=rtol)
```
From the docs it's not obvious to me how `atol` and `rtol` are interpreted as threshold. Could you add a paragraph to the docstring that says that we check the operator times it's adjoint is the identity, up to the tolerance? E.g. something like
```python
"""Return True if the operator represents a unitary matrix.

This method checks whether the operator composed with it's 
adjoint equals the identity, up to the provided tolerance.
```

littlebullGit replied:
doc string update

Cryoris commented on test/python/quantum_info/operators/symplectic/test_sparse_pauli_op.py:
```diff
@@ -1342,6 +1342,29 @@ def test_apply_layout_zero_qubit(self, layout):
             self.assertEqual(SparsePauliOp.from_list([("IIIII", 1), ("IIIII", 2)]), res)
 
+    def test_is_unitary_tolerance(self):
+        """Test that is_unitary respects the input tolerance values."""
```
It would be nice to link the original issue here to show where these values are coming from. Typically we do
```suggestion
        """Test that is_unitary respects the input tolerance values.
        
        Regression test of #14107.
        """
```

Cryoris reviewed (commented):
Thanks for taking a look at this. Could you also add a bugfix release note to this? You can check the contribution guidelines (or the `releasenotes/notes` folder) for some guidance on this

littlebullGit committed:
Fix SparsePauliOp.is_unitary() to respect input tolerance values

This commit fixes issue #14107 by ensuring that the tolerance values passed to
is_unitary() are properly used when checking if the operator is unitary. The
tolerances are now correctly passed through to the internal operations.

* Update docstring to better explain how tolerance parameters are used
* Add test case to verify tolerance behavior
* Add release note for the bug fix

littlebullGit commented:
applied the PR comments and rebased on upstream main. please verify @Cryoris

Cryoris commented:
It looks like something went wrong in your rebase (you're added as co-author now on _all_ commits from your rebase), we should only see commits associated with this PR here.

littlebullGit commented:
> It looks like something went wrong in your rebase (you're added as co-author now on _all_ commits from your rebase), we should only see commits associated with this PR here.

I have create a new local branch from upstream main, cherry pick my commit and pushed again. please check @Cryoris

Cryoris commented on releasenotes/notes/fix-sparse-pauli-op-is-unitary-tolerance-28249b2d55f38284.yaml:
```diff
@@ -0,0 +1,12 @@
+    Fixed an issue where :meth:`~qiskit.quantum_info.SparsePauliOp.is_unitary`
+    was not properly respecting the input tolerance values when checking if an
+    operator is unitary. The method now correctly uses the provided `atol` and
+    `rtol` parameters when simplifying the operator and checking if it equals
```
Single quotes for sphinx annotations and double quotes for code-font 🙂 
```suggestion
    operator is unitary. The method now correctly uses the provided ``atol`` and
    ``rtol`` parameters when simplifying the operator and checking if it equals
```

Cryoris commented on releasenotes/notes/fix-sparse-pauli-op-is-unitary-tolerance-28249b2d55f38284.yaml:
```diff
@@ -0,0 +1,12 @@
+    was not properly respecting the input tolerance values when checking if an
+    operator is unitary. The method now correctly uses the provided `atol` and
+    `rtol` parameters when simplifying the operator and checking if it equals
+    the identity. This fixes :issue:`14107`.
```
I don't think `:issue:` is a directive we use, you can just use 
```
`#14107 <link to issue>`__
```
instead

littlebullGit replied:
Both implemented.

Cryoris reviewed (commented):
Yeah that looks better! Some minor comments below

littlebullGit committed:
update based on  Cryoris' comments

Cryoris commented on releasenotes/notes/fix-sparse-pauli-op-is-unitary-tolerance-28249b2d55f38284.yaml:
```diff
@@ -0,0 +1,12 @@
+    This change ensures that the tolerance values passed to `is_unitary()` are
+    properly propagated to the internal simplification and comparison operations,
+    making the method more consistent with user expectations and other similar
+    methods in Qiskit.
```
```suggestion

```

Cryoris commented:
@littlebullGit one of the files is not correctly formatted, you can resolve this by running the black formatter (e.g. `make black` in the root directory) 🙂

littlebullGit committed:
format test file with black and update release notes based on comments
